### PR TITLE
chore: Integration Account - Provide script to upload maps.

### LINKED ISSUE #161: Provide script to upload maps (transformations) into an Integration Account
**Is your feature request related to a problem? Please describe.**
In order to be able to upload transformations (XSLT, XSLT2.0, XSLT3.0 or liquid) into an Integration Account, this either has to be done manually or some custom scripts.
Since there is already [a script](https://codit.visualstudio.com/invictus%20Templates/_git/invictus-templates?path=%2Fdeploy%2FIntegration-DeployIntegrationAccountArtifacts.ps1) that offers this functionality (in a bigger picture) we can extract the logic to upload maps into a separate script. 
This allows users to either only upload maps, or run the larger command to upload all available Integration Account artifacts.

The given script uses the REST API since no scripts were available yet at the time, or it had severe limitations.
To be checked whether this can be replaced by the following commands:
- [New-AzIntegrationAccountMap](https://docs.microsoft.com/en-us/powershell/module/az.logicapp/new-azintegrationaccountmap?view=azps-5.8.0)
- [Set-AzIntegrationAccountMap](https://docs.microsoft.com/en-us/powershell/module/az.logicapp/set-azintegrationaccountmap?view=azps-5.8.0)

**Describe the solution you'd like**
The script should accept the following parameters:
- IntegrationAccountName
  The name of the Integration Account to be targetted.
- IntegrationAccountResourceGroup
  The name of the resource group containing the Integration Account
- MapsFolder
  Base folder containing all maps to be uploaded.
- ArtifactsPrefix (optional)
  When present, add this prefix to the name of the maps before uploading (in case of sharing an Integration Account across multiple environments in order to save costs.

The script should loop through the provided folder, given the following prerequisites regarding folder-structure:
- /Maps/Xslt
- /Maps/Xslt2.0
- /Maps/Xslt3.0
- /Maps/Liquid
The MapType is to be set during the upload process.

**Additional context**
This is part of a number of scripts to upload artifacts into an Azure Integration Account, in order to replace the `Integration-DeployIntegrationAccountArtifacts.ps1` from the Invictus Templates.

### PR DESCRIPTION
**Pull request by mbraekman**
This PR closes #161

In order to align with other scripts, the following **parameters** mentioned in the issue have been renamed:
- IntegrationAccountName --> Name
- IntegrationAccountResourceGroup --> ResourceGroup

The following **parameters** have been added:
- MapsFolder: allows upserting multiple maps
- MapFilePath: allows upserting a single map
- MapType: allows specifying the map-type. (default to 'Xslt')
- ArtifactsPrefix: allows adding a prefix to the names before uploading the maps.
- RemoveFileExtensions: switch indicating whether to remove the file extension.

**Module:**
Arcus.Scripting.IntegrationAccount.

**Documentation:**
The module-specific page (Integration Account) has been updated in the preview folder.

**Testing:**
Both the _Integration_ and _Unit_ testing modules have been extended for this function.

### CODE REVIEW

File: docs/preview/features/powershell/azure-integration-account.md [hunk 1]
```diff
@@ -85,5 +86,74 @@ PS> Set-AzIntegrationAccountSchemas -ResourceGroupName 'my-resource-group' -Name
...
+| ---------------------- | --------- | -------------------------------------------------------------------------------------------------------------------------------------- |
+| `ResourceGroupName`    | yes       | The name of the Azure resource group where the Azure Integration Account is located.                                                   |
+| `Name`                 | yes       | The name of the Azure Integration Account into which the maps are to be uploaded/updated.                                              |
+| `MapFilePath`          | no        | The full path of a map that should be uploaded/updated. (Mandatory if MapsFolder has not been specified).                              |
+| `MapsFolder`           | no        | The path to a directory containing all maps that should be uploaded/updated. (Mandatory if MapFilePath has not been specified).        |
+| `MapType`              | no        | The type of map to be created, default to 'Xslt'. See possible values [here](https://docs.microsoft.com/en-us/powershell/module/az.logicapp/get-azintegrationaccountmap?view=azps-6.2.1#parameters).  |
```
Comment (stijnmoreels) [inline docs/preview/features/powershell/azure-integration-account.md:99]: Maybe these should have something else in the `Mandatory` column? Like `yes (when ... is not specified)` or `sometimes` or...

File: src/Arcus.Scripting.IntegrationAccount/Scripts/Set-AzIntegrationAccountMaps.ps1 [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            Write-Verbose "No map '$mapName' could not be found in Azure Integration Account '$Name'."
+        }
+        else {
+            throw $_.Exception
+        }
+    }
+        
+    try {
+        if ($existingMap -eq $null) {
+            # Create the map
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.IntegrationAccount/Scripts/Set-AzIntegrationAccountMaps.ps1:47]:
```suggestion
```

Nit, not sure if this comment has any added-value.

File: src/Arcus.Scripting.IntegrationAccount/Scripts/Set-AzIntegrationAccountMaps.ps1 [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            Write-Verbose "Creating map '$mapName' in Azure Integration Account '$Name'"
+            $createdMap = New-AzIntegrationAccountMap -ResourceGroupName $ResourceGroupName -Name $Name -MapName $mapName -MapType $MapType -MapFilePath $Map.FullName -ErrorAction Stop
+            Write-Verbose ($createdMap | Format-List -Force | Out-String)
+        }
+        else {
+            # Update the map
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.IntegrationAccount/Scripts/Set-AzIntegrationAccountMaps.ps1:53]:
```suggestion
```

Nit, not sure if this comment has any added-value.

File: src/Arcus.Scripting.IntegrationAccount/Scripts/Set-AzIntegrationAccountMaps.ps1 [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    $mapName = $Map.Name
+    if ($RemoveFileExtensions) {
+        $mapName = $Map.BaseName
+    }
+    if ($ArtifactsPrefix -ne '') {
+        $mapName = $ArtifactsPrefix + $mapName
+    }
+    Write-Host "Uploading map '$mapName' into the Integration Account '$Name'."
+
+    ## Check if the map already exists
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.IntegrationAccount/Scripts/Set-AzIntegrationAccountMaps.ps1:30]:
```suggestion
```

Nit, not sure if this comment has any added-value.

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 [hunk 5]
```diff
@@ -175,7 +201,8 @@ InModuleScope Arcus.Scripting.IntegrationAccount {
                         
                         $actual = Get-AzIntegrationAccountSchema -ResourceGroupName $resourceGroupName -Name $integrationAccountName -SchemaName $expectedSchemaName
                         $actual | Should -Not -BeNullOrEmpty
-                        $actual.CreatedTime.ToString("yyyy-MM-ddTHH:mm:ss") | Should -Be $actual.ChangedTime.ToString("yyyy-MM-ddTHH:mm:ss")
+                        $actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -BeIn ($actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss"), $actual.ChangedTime.ToUniversalTime().AddSeconds(-1).ToString("yyyy-MM-ddTHH:mm:ss"))
+                        #$actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -Be $actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss")
                         $actual.CreatedTime | Should -BeGreaterOrEqual $executionDateTime
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:206]:
```suggestion
                        $actual.CreatedTime | Should -BeGreaterOrEqual $executionDateTime
```

Can be removed?

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 [hunk 6]
```diff
@@ -187,7 +214,215 @@ InModuleScope Arcus.Scripting.IntegrationAccount {
...
+
+                try {
+                    # Act
+                    Set-AzIntegrationAccountMaps -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapFilePath $map.FullName
+
+                    # Assert
+                    $actual = Get-AzIntegrationAccountMap -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapName $expectedMapName
+                    $actual | Should -Not -BeNullOrEmpty 
+                    $actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -BeIn @($actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss"), $actual.ChangedTime.ToUniversalTime().AddSeconds(-1).ToString("yyyy-MM-ddTHH:mm:ss"))
+                    #$actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -Be $actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss")
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:252]:
```suggestion
```

Can be removed?

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 [hunk 7]
```diff
@@ -187,7 +214,215 @@ InModuleScope Arcus.Scripting.IntegrationAccount {
...
+
+                try {
+                    # Act
+                    Set-AzIntegrationAccountMaps -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapFilePath $map.FullName
+
+                    # Assert
+                    $actual = Get-AzIntegrationAccountMap -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapName $expectedMapName
+                    $actual | Should -Not -BeNullOrEmpty
+                    $actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -BeIn ($existingMap.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss"), $existingMap.ChangedTime.ToUniversalTime().AddSeconds(-1).ToString("yyyy-MM-ddTHH:mm:ss"))
+                    #$actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -Be $existingMap.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss")
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:278]:
```suggestion
```

Can be removed?

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 [hunk 8]
```diff
@@ -187,7 +214,215 @@ InModuleScope Arcus.Scripting.IntegrationAccount {
...
+
+                try {
+                    # Act
+                    Set-AzIntegrationAccountMaps -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapFilePath $map.FullName -RemoveFileExtensions
+
+                    # Assert
+                    $actual = Get-AzIntegrationAccountMap -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapName $expectedMapName
+                    $actual | Should -Not -BeNullOrEmpty
+                    $actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -BeIn ($actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss"), $actual.ChangedTime.ToUniversalTime().AddSeconds(-1).ToString("yyyy-MM-ddTHH:mm:ss"))
+                    #$actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -Be $actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss")
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:303]:
```suggestion
```

Can be removed?

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 [hunk 9]
```diff
@@ -187,7 +214,215 @@ InModuleScope Arcus.Scripting.IntegrationAccount {
...
+
+                try {
+                    # Act
+                    Set-AzIntegrationAccountMaps -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapFilePath $map.FullName -ArtifactsPrefix $artifactsPrefix -RemoveFileExtensions
+
+                    # Assert
+                    $actual = Get-AzIntegrationAccountMap -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapName $expectedMapName
+                    $actual | Should -Not -BeNullOrEmpty
+                    $actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -BeIn ($actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss"), $actual.ChangedTime.ToUniversalTime().AddSeconds(-1).ToString("yyyy-MM-ddTHH:mm:ss"))
+                    #$actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -Be $actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss")
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:328]:
```suggestion
```

Can be removed?

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 [hunk 10]
```diff
@@ -187,7 +214,215 @@ InModuleScope Arcus.Scripting.IntegrationAccount {
...
+                    Set-AzIntegrationAccountMaps -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapsFolder $mapsFolder
+
+                    # Assert
+                    foreach ($map in Get-ChildItem($mapsFolder) -File) {
+                        $expectedMapName = $map.Name
+                        
+                        $actual = Get-AzIntegrationAccountMap -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapName $expectedMapName
+                        $actual | Should -Not -BeNullOrEmpty
+                        $actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -BeIn ($actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss"), $actual.ChangedTime.ToUniversalTime().AddSeconds(-1).ToString("yyyy-MM-ddTHH:mm:ss"))
+                        #$actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -Be $actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss")
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:353]:
```suggestion
```

Can be removed?

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 [hunk 11]
```diff
@@ -187,7 +214,215 @@ InModuleScope Arcus.Scripting.IntegrationAccount {
...
+                    Set-AzIntegrationAccountMaps -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapsFolder $mapsFolder -RemoveFileExtensions
+
+                    # Assert
+                    foreach ($map in Get-ChildItem($mapsFolder) -File) {
+                        $expectedMapName = $map.BaseName
+                        
+                        $actual = Get-AzIntegrationAccountMap -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapName $expectedMapName
+                        $actual | Should -Not -BeNullOrEmpty
+                        $actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -BeIn ($actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss"), $actual.ChangedTime.ToUniversalTime().AddSeconds(-1).ToString("yyyy-MM-ddTHH:mm:ss"))
+                        #$actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -Be $actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss")
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:383]:
```suggestion
```

Am I repeating myself? 😄

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 [hunk 12]
```diff
@@ -187,7 +214,215 @@ InModuleScope Arcus.Scripting.IntegrationAccount {
...
+                    Set-AzIntegrationAccountMaps -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapsFolder $mapsFolder -ArtifactsPrefix $artifactsPrefix -RemoveFileExtensions
+
+                    # Assert
+                    foreach ($map in Get-ChildItem($mapsFolder) -File) {
+                        $expectedMapName = $artifactsPrefix + $map.BaseName
+                        
+                        $actual = Get-AzIntegrationAccountMap -ResourceGroupName $resourceGroupName -Name $integrationAccountName -MapName $expectedMapName
+                        $actual | Should -Not -BeNullOrEmpty
+                        $actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -BeIn ($actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss"), $actual.ChangedTime.ToUniversalTime().AddSeconds(-1).ToString("yyyy-MM-ddTHH:mm:ss"))
+                        #$actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -Be $actual.ChangedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss")
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:414]:
```suggestion
```

Probably can be removed?

**Review by stijnmoreels — APPROVED** [review]
Great set of tests, great doc section! 🥇

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 [hunk 13]
```diff
@@ -45,21 +45,47 @@ InModuleScope Arcus.Scripting.IntegrationAccount {
...
+                $existingSchema = New-AzIntegrationAccountSchema -ResourceGroupName $resourceGroupName -Name $integrationAccountName -SchemaName $expectedSchemaName -SchemaFilePath $schema.FullName
+
+                try {
+                    # Act
+                    Set-AzIntegrationAccountSchemas -ResourceGroupName $resourceGroupName -Name $integrationAccountName -SchemaFilePath $schema.FullName
+
+                    # Assert
+                    $actual = Get-AzIntegrationAccountSchema -ResourceGroupName $resourceGroupName -Name $integrationAccountName -SchemaName $expectedSchemaName
+                    $actual | Should -Not -BeNullOrEmpty
+                    $actual.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss") | Should -BeIn ($existingSchema.CreatedTime.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ss"), $existingSchema.CreatedTime.ToUniversalTime().AddSeconds(-1).ToString("yyyy-MM-ddTHH:mm:ss"))
```
Comment (stijnmoreels) [inline src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:73]: Always a ___ in the ___ to assert on dates.

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 (hunk 5 above)
Comment (mbraekman) [reply src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:206]: Argh, yes..
Should've checked and removed these comments before committing 😅

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 (hunk 11 above)
Comment (mbraekman) [reply src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:383]: My bad 😄 
The worst thing about this is, while I was turning those lines in comments, I was thinking I shouldn't forget to actually remove them...

File: src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1 (hunk 13 above)
Comment (mbraekman) [reply src/Arcus.Scripting.Tests.Integration/Arcus.Scripting.IntegrationAccount.tests.ps1:73]: 💯

File: docs/preview/features/powershell/azure-integration-account.md (hunk 1 above)
Comment (mbraekman) [reply docs/preview/features/powershell/azure-integration-account.md:99]: Good point!
I'll change this to "conditional" and set the actual condition in _italic_ in the description-column.

**Comment by mbraekman** [comment]
> In terms of the failing tests:
> 
> ```
> 'The content of map 'BankTransfer_CSV-to-BankTransfer_Canonical' of type 'Xslt' must be a valid XML._
> ```
> 
> Seems that the test files are missing something? They're UTF-16. That's correct?

@stijnmoreels, the UTF-16 encoding is actually correct, it seems like there was an issue in the actual structure of the XML. Which has now been resolved ;)

**Pull request merged**